### ISSUE TITLE
complex source-point Gaussian beam

### ISSUE THREAD

**Issue by stevengj** [NanoComp/meep#1094]
The current [tutorial](https://meep.readthedocs.io/en/latest/FAQ/#how-do-i-create-a-focused-beam-with-a-gaussian-envelope) on methods to create a Gaussian incident beam is a bit primitive (paraxial approx, source term located at beam waist only).  It would be nice to implement a more accurate and general Gaussian-beam source using the complex-source-point method.

SCUFF already has an implementation on this in [GaussianBeam.cc](https://github.com/HomerReid/scuff-em/blob/9c6d0cb7695463af803dee8d04cdae939740cdcc/libs/libIncField/GaussianBeam.cc) that we could adopt for Meep.  Further references and variations can be found in paragraph 5 of [this paper](http://math.mit.edu/~stevenj/papers/WongKa14.pdf).

Basically, the complex-source Gaussian beam formulation gives an exact analytical solution of Maxwell's equations that corresponds to a Gaussian beam.  Given this solution, we can then apply the equivalence principle to obtain corresponding current sources.

(Since this is a frequency-domain formulation, of course, it will only be an accurate Gaussian beam within a relatively narrow bandwidth.)

**Comment by stevengj** [comment]
In cylindrical coordinates: https://www.osapublishing.org/aop/abstract.cfm?URI=aop-1-1-1

**Comment by oskooi** [comment]
Is this a duplicate of #711?

**Comment by stevengj** [comment]
Whoops, right.
